### PR TITLE
avoid ACS linear contract ID lookups in Trigger library

### LINKED ISSUE #7112: Use GenMap in DAML Trigger API
Currently, triggers do a bunch of linear searches because we cannot have maps indexed by contract ids and type-indexed maps. This should be changed to use maps internally for DAML-LF 1.dev.

This will require CPP for now.

### PR DESCRIPTION
**Pull request by S11001001**
Adds another Map layer to the trigger ACS, which is now `Map TemplateTypeRep (Map AnyContractId AnyTemplate)`.

Fixes #7112, followup to #7365.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/daml/Daml/Trigger/Internal.daml [hunk 1]
```diff
@@ -39,8 +39,11 @@ data ACS = ACS
   { activeContracts :
 #ifdef DAML_GENMAP && DAML_GENERIC_COMPARISON
       GMap.Map TemplateTypeRep
+         (GMap.Map
+#else
+      [] ((,)
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/Internal.daml:44]: This seems a bit too code golfy. I think just moving `AnyContractId AnyTemplate` in both branches would be easier to parse`.

File: triggers/daml/Daml/Trigger/Internal.daml [hunk 2]
```diff
@@ -56,8 +59,8 @@ addCommands : Map CommandId [Command] -> Commands -> Map CommandId [Command]
 addCommands m (Commands cid cmds) = Map.insert cid cmds m
 
 #ifdef DAML_GENMAP && DAML_GENERIC_COMPARISON
-adjust : forall k v. Ord k => k -> (Optional v -> Optional v) -> GMap.Map k v -> GMap.Map k v
-adjust k f m =
+alter : forall k v. Ord k => k -> (Optional v -> Optional v) -> GMap.Map k v -> GMap.Map k v
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/Internal.daml:62]: In Haskell, the order of key and the modifier function is flipped. Not an issue here but once we move this to `daml-stdlib`,  we should probably try to match that unless we have a good reason not to do so.

File: triggers/daml/Daml/Trigger/Internal.daml [hunk 3]
```diff
@@ -67,40 +70,39 @@ adjust k f m =
...
+  where addct = Some . GMap.insert cid tpl . fromOptional mempty
 #else
 insertTpl cid tpl acs = acs { activeContracts = (cid, tpl) :: acs.activeContracts }
 #endif
 
 #ifdef DAML_GENMAP && DAML_GENERIC_COMPARISON
-groupActiveContracts : [(AnyContractId, AnyTemplate)] -> GMap.Map TemplateTypeRep [(AnyContractId, AnyTemplate)]
-groupActiveContracts = foldr (\v@(cid, _) -> adjust cid.templateId (addct v)) GMap.empty
-  where addct v ol = Some $ v :: fromOptional [] ol
+groupActiveContracts : [(AnyContractId, AnyTemplate)]
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/Internal.daml:80]: nit: I’m generally in favor of not making indentation depend on identifier length so a line-break before the start of the type signature.

**Review by cocreature — APPROVED** [review]
Nice, thank you!

File: triggers/daml/Daml/Trigger/Internal.daml (hunk 1 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/Internal.daml]: Sure. c22f62e8cc

File: triggers/daml/Daml/Trigger/Internal.daml (hunk 2 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/Internal.daml]: Might as well do it now. 4cba20c0dc

File: triggers/daml/Daml/Trigger/Internal.daml (hunk 3 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/Internal.daml]: No problem. e58ffae6ec

**Pull request merged**